Tools/ardupilotwaf/toolchain.py

```diff
@@ -24,14 +25,71 @@ def configure(cfg):
     OBJCOPY='objcopy',
 )
 
+def find_realexec_path(cfg, filename, path_list=[]):
+    if not filename:
+        return ''
+
+    if not path_list:
+        path_list = cfg.environ.get('PATH','').split(os.pathsep)
+
+    for dir in path_list:
+        path = os.path.abspath(os.path.expanduser(os.path.join(dir, filename)))
+
+        if os.path.isfile(path):
+            if os.path.islink(path):
+                realpath = os.path.realpath(path)
+
+                if filename not in os.path.basename(realpath):
+                    continue
+                else:
+                    return os.path.dirname(realpath)
+
+            else:
+                return os.path.dirname(path)
+
+    cfg.fatal('Could not find real exec path to %s in path_list %s:' % (filename, path_list))
+
 def configure(cfg):
-    if not cfg.env.TOOLCHAIN:
-        cfg.env.TOOLCHAIN = 'native'
-        prefix = ''
-    else:
-        cfg.env.TOOLCHAIN = Utils.to_list(cfg.env.TOOLCHAIN)[0]
-        cfg.msg('Using toolchain prefix', cfg.env.TOOLCHAIN)
-        prefix = cfg.env.TOOLCHAIN + '-'
-
-    for k in suffixes:
-        cfg.env.append_value(k, prefix + suffixes[k])
+    toolchain = boards.get_board(cfg.env.BOARD).toolchain
+
+    if toolchain != 'native':
+        cfg.msg('Using toolchain prefix', toolchain)
+        prefix = toolchain + '-'
+
+        c_compiler = cfg.options.check_c_compiler or 'gcc'
+        cxx_compiler = cfg.options.check_cxx_compiler or 'g++'
+
+        toolchain_path = os.path.abspath(os.path.join(find_realexec_path(cfg, prefix + suffixes['AR']), '..'))
+        sysroot = cfg.cmd_and_log([prefix + 'gcc', '--print-sysroot'], quiet=Context.BOTH)[:-1]
+
+        cfg.msg('Using toolchain path', toolchain_path)
+        
+        clang_flags = [
+            '--target=' + toolchain,
+            '--gcc-toolchain=' + toolchain_path,
+            '--sysroot=' + sysroot,
+            '-B' + os.path.join(toolchain_path, 'bin')
+        ]
+
+        if 'gcc' == c_compiler:
+            cfg.env['CC'] = [ prefix + 'gcc' ]
+
+        elif 'clang' == c_compiler:
+            cfg.env['CC'] = [ c_compiler ]
+            cfg.env['AR'] = [ prefix + suffixes['AR'] ]
+            cfg.env.CFLAGS += clang_flags
+
+        if 'g++' == cxx_compiler:
+            cfg.env['CXX'] = [ prefix + 'g++' ]
+
+        elif 'clang++' == cxx_compiler:
+            cfg.env['CXX'] = [ cxx_compiler ]
+            cfg.env['AR'] = [ prefix + suffixes['AR'] ]
+            cfg.env.CXXFLAGS += clang_flags
+
+        if 'gcc' == c_compiler or 'g++' == cxx_compiler:
+            for k in suffixes:
+                cfg.env[k] = [ prefix + suffixes[k] ]
+                
+        if 'clang' == c_compiler or 'clang++' == cxx_compiler:
+            cfg.env.LINKFLAGS += clang_flags
```

Comment:
[Comment for the whole commit]

@OXINARF I still think there's a better way to do this... I think we could try "hijacking" compiler_c and compiler_cxx tools. Do you mind if I try an alternative patch on top on what you've done so far? The hijacking would be changing   values in `waflib.Tools.compiler_cxx.cxx_compiler` and also for compiler_c.

Then we could discuss about both approaches.